### ISSUE TITLE
Third-party config (hx.react/defnc, compojure, etc)

### ISSUE THREAD

**Issue by orestis** [borkdude/clj-kondo#559]
hx is a React wrapper. It works by defining its own `defnc` macro to pre-process hiccup.

The linting rule is just this:
```{:lint-as {hx.react/defnc clojure.core/defn}}```

Would be nice to get this built in clj-kondo :)

**Comment by orestis** [comment]
I think you explained on Slack already, but I didn’t get what is the technical holdup to allow library authors to expose linting rules as part of macro metadata. If people have to pre-warm the cache to get all the symbols for the library anyway, then that needs to happen before linting the consumer namespace. 

But OTOH it’s the end of the day and I have no idea how things work under the hood 😅

**Comment by borkdude** [comment]
@orestis Some reasons:

My assumption is that a lot of people don't really lint all their libraries and just fire up their editor and start working. Meanwhile clj-kondo save information about every edited file in the cache. So when people define a function, the arity is saved into the cache. When they use a function somewhere else, the cache is consulted.

Requiring people to lint from the command line to get the config for a third-party macro is an extra step that people might or might not do.

Also, when config is accumulating from linted sources, configuration issues becomes harder to debug. When invalid config ends up in the cache, it can derail clj-kondo (because of unexpected values, etc). And it becomes a lot harder to help people with issues they are seeing.

**Comment by borkdude** [comment]
@orestis

I wrote down some of the principles I've been using to build clj-kondo here:

https://github.com/borkdude/clj-kondo/blob/master/doc/dev.md#design-principles

I hope it adds to the understanding of trade-offs.

**Comment by borkdude** [comment]
TODO:

- add config for build.boot
- add config for clara.rules #576 
- add config for compojure

**Comment by geraldodev** [comment]
support for defsc macro from fulcro would be nice 

http://book.fulcrologic.com/#_the_defsc_macro_2

**Comment by borkdude** [comment]
@geraldodev Does `{:lint-as {com.fulcrologic.fulcro.components/defsc clojure.core/defn}}` work?

**Comment by geraldodev** [comment]
@borkdude  It worked. Thank you for the tool. Very useful and fast !

**Comment by borkdude** [comment]
I think having a configuration section with lots and lots of examples and reference configs for third party libraries is better than having this built into clj-kondo.

1) people will learn how to use the configuration options when visiting that page
2) clj-kondo won't become slower as the result of a growing built-in config

**Comment by borkdude** [comment]
Related: https://github.com/borkdude/clj-kondo/issues/253

**Comment by borkdude** [comment]
Another idea:

Libraries could at startup emit some piece of EDN config to a designated directory. e.g. .clj-kondo/libs/clara.rules/config.edn.

The benefit of this would be that only config for the relevant libraries in one project would end up there.

Extra attention would be needed to provide backward compatibility if older libraries keep emitting old config formats. And what about wrong formats even? Extra attention is needed for exception handling.

**Comment by borkdude** [comment]
Clj-kondo will probably not pull the `.clj-kondo/config.edn` since that's not on the classpath, but if libs stick to something like src/clj-kondo/config.edn for stuff they want to export, that could work
And then libs can make a separation between stuff they want to export and config they are using themselves.

**Comment by kwrooijen** [comment]
That means that `src/clj-kondo/config.edn` will be in the user's path, right? Currently Duct has the convention to use `config.edn` as its "main" configuration endpoint (by default). I'm not sure how the priority is determined in classpaths (first in first out?). Maybe dependency classpaths have lower priority?

Anyway, having to think of every tool / framework is probably pointless. But I'm just letting you know that this could potentially break Duct apps.

**Comment by borkdude** [comment]
@kwrooijen Thanks for sharing. Can you point me to the relevant documentation in duct?

**Comment by kwrooijen** [comment]
Duct's docs are a bit lacking, but this is probably good enough: https://github.com/duct-framework/docs/blob/master/GUIDE.rst#configuration

**Comment by borkdude** [comment]
Note, if libraries are going to export config on the classpath they'll have to do it in a directory that is unique by their org+lib name, else you'll get classpath conflicts:

E.g.
```
src/acme/coollib/clj-kondo/config.edn
```
but then it becomes hard for clj-kondo to detect where that config is.

So maybe:
```
src/clj-kondo/acme/coollib/config.edn
src/clj-kondo/acme/coollib/hooks/acme/coollib/weird_macroc.clj
```
is a good convention?

**Comment by rschmukler** [comment]
The only thing that feels weird about that to me is that typically in most projects `src/acme/coollib` ends up sort of functioning as a "project-root". I wonder if perhaps `clj-kondo` could match the namespace of `*.clj-kondo.*`. The downside here is that if any library has a `clj-kondo` path that isn't actually `clj-kondo` related, that could break. Also, will both of these end up meaning that `clj-kondo` can't be used to lint itself?

Another option might be something like an attribute on the namespace. Eg:

```clojure
(ns acme.coollib.clj-kondo
  {:clj-kondo true})

(defn config
  "Return some edn"
  []
  {:lint-as 'acme.coollib.core/defn defn})
```

No real opinions yet, just throwing more things out

**Comment by Pancia** [comment]
Copying message & idea from slack:

> borkdude  2:12 PM
> we might do better here by providing a classpath option, so you can put your hooks in a github project and use clojure to resolve that library into a classpath

Original question:
> adambros  2:02 PM
> What are my options if I want to refer in my config.edn hooks to a rewrite that i want provided as a library or just in my home dir? can I refer to a hooks that is outside of the .clj-kondo directory?
> I'm using vim / ALE and clojure-lsp, but for ale i would have to write a whole new linter file to pass it --config, and clojure-lsp just doesnt support that currently... (edited)

**Comment by borkdude** [comment]
About the idea @Pancia mentioned, I think #992 might be a nice place for further discussions.

**Comment by borkdude** [comment]
Further discussion in #992, as that brings together other related issues as well.

**Comment by borkdude** [comment]
I just merged an implementation of exporting and importing config via the classpath to master. Docs:

https://github.com/borkdude/clj-kondo/blob/master/doc/config.md#Exporting-and-importing-configuration

**Comment by rschmukler** [comment]
@borkdude this is perfect! Looking forward to bundling up some configs in my libraries! Thanks as always for your efforts.
